testflinger_agent/__init__.py

```diff
@@ -26,6 +31,87 @@
 logger = logging.getLogger(__name__)
 
 
+class ReqBufferTimer(Timer):
+    """Requests buffer flush"""
+
+    def run(self):
+        """Loop timer"""
+        while not self.finished.wait(self.interval):
+            self.function(*self.args, **self.kwargs)
+
+
+class ReqBufferHandler(logging.Handler):
+    """Requests logging handler"""
+
+    def __init__(self, agent, server):
+        super().__init__()
+        uri = urljoin(server, "/v1/agents/data/")
+        self.url = urljoin(uri, agent)
+        self.qdepth = 100  # messages
+        self.buffer = deque([], maxlen=self.qdepth)
+        self.reqbuf_timer = None
+        self.reqbuf_interval = 10.0  # seconds
+        self._start_rb_timer()
+        # reuse socket
+        self.session = self._requests_retry()
+
+    def _requests_retry(self, retries=3):
+        """Retry api server"""
+        session = requests.Session()
+        retry = Retry(
+            total=retries,
+            read=retries,
+            connect=retries,
+            backoff_factor=0.3,
+            status_forcelist=(500, 502, 503, 504),
+        )
+        adapter = HTTPAdapter(max_retries=retry)
+        session.mount("http://", adapter)
+        session.mount("https://", adapter)
+        return session
+
+    def _start_rb_timer(self):
+        """Periodically check and send buffer"""
+        self.reqbuf_timer = ReqBufferTimer(self.reqbuf_interval, self.flush)
+        # terminate timer on exit
+        self.reqbuf_timer.daemon = True
+        self.reqbuf_timer.start()
+
+    def emit(self, record):
+        """Write logging events to buffer"""
+        if len(self.buffer) >= self.qdepth:
+            self.buffer.popleft()
+
+        self.buffer.append(record)
+
+    def flush(self):
+        """Flush and post buffer"""
+        try:
+            for record in self.buffer:
+                self.session.post(
+                    url=self.url, json=self.format(record), timeout=3
+                )
+        except Exception as e:
+            logger.exception(e)
```

Comment:
Ah, ok I think this is what I was seeing. This is going to be a VERY common case until we deploy the new server code, so we probably want to just put out a logger.debug() or something instead, and warn that we couldn't flush the agent logs to the server, and not a logger.exception, which is going to generate a LOT of spam in all of the agent logs for a very long time

Also, please avoid "except Exception...". We're trying to migrate towards pylint for everything and it will definitely complain about that. Catch specific exceptions whenever possible, or if it *really* needs a broad exception, then find out all the exception types it could get and see if they share a parent class like OSError or IOError

Reply:
Yes, I agree with specific exceptions, and use pylint myself. Saw basic broad exception handling elsewhere, so I wasn't sure if this was a desired pattern for simplicity.